### DISCUSSION TITLE
Proposed `Dataset` API changes

### DISCUSSION THREAD

**Discussion by namedgraph** [RDFLib/rdflib#2591] [Development]
The `Dataset` is quite weird and assumes that standalone `Graph`s have identifiers, which will be phased out (#2537). For example, adding a named graph to a `Dataset` looks like this:
```python
g = Graph(identifier=URIRef("urn:named_graph"))
g.add((..., ... ,...))
...
d = Dataset()
d.add_graph(g)
```
Moreover, `Dataset` uses the term `context` when referring to named graphs. I think it should be phased out as well.
If in doubt, I suggest just copying Jena's [`Dataset` API](https://jena.apache.org/documentation/javadoc/arq/org.apache.jena.arq/org/apache/jena/query/Dataset.html).

My suggestions for `Dataset`:
* add `add_named_graph(uri: IdentifiedNode, graph: Graph)` method
* add `has_named_graph(uri: IdentifiedNode)` method
* add `remove_named_graph(uri: IdentifiedNode)` method
* add `replace_named_graph(uri: IdentifiedNode, graph: Graph))` method
* add `graphs()` method as an alias for `contexts()`
* add `default_graph` property as an alias for `default_context`
* add `get_named_graph` as an alias for `get_graph`
* deprecate `graph(graph)` method
* deprecate `remove_graph(graph)` method
* deprecate `contexts()` method

Using `IdentifiedNode` as a [super-interface for `URIRef` and `BNode`](https://rdflib.readthedocs.io/en/stable/rdf_terms.html) (since both are allowed as graph names in RDF 1.1).

The above example would become something like this after these changes:
```python
g = Graph()
g.add((..., ... ,...))
...
d = Dataset()
d.add_named_graph(URIRef("urn:named_graph"), g)
```

**Comment by namedgraph** [comment]
Also aligns with [#2446](https://github.com/RDFLib/rdflib/discussions/2446) and #2407

**Comment by namedgraph** [comment]
It also looks like a simple `Graph.add(self, g: Graph)` method is missing 🤷‍♂️  There's only `__add__`.

**Comment by nicholascar** [comment]
I support all of the proposals in this discussion. This has been a long-time coming - we've noticed these things for years - but have never done anything about these and they still hurt u - @edmondchuc is battling with datasets in a current project.

I suggest we also remove the `ConjunctiveGraph` class and fold any differences it has with `Dataset` into `Dataset` constructor parameters.

**Comment by namedgraph** [comment]
I don't think `ConjunctiveGraph` has to go -- as I understand, it provides a "union graph" which most triplestores support. But `Dataset` probably does not need to extend it.

**Comment by recalcitrantsupplant** [comment]
I've wrote down my thoughts on what expected interfaces are in a pseudo python/rdflib format:
- without reference to the current implementation. As such I don't expect what I've written below would be backwards compatible - I'd think it should be changed for this reason.
- without reading other's suggestions  

Hopefully it's a coherent perspective; it may take some effort to reconcile / integrate with others'. Will have a go at this next.

## Minimal class definitions
Only enough to illustrate the thinking/scenarios
Graph:
```python
class GraphType(Enum):
    DEFAULT = "default"
    NAMED = "named"

class Graph:
    def __init__(
        self,
        identifier: URIRef | None = None,
        graph_type: GraphType | None = None,
    ):
        pass
```
Dataset:
```python
class Dataset:
    def __init__(self):
        pass

    def quads(
        self,
        context: GraphType | URIRef | list[GraphType | URIRef] | None = None,
    ):
        pass

    def triples(
        self,
        context: GraphType | URIRef | list[GraphType | URIRef] | None = None,
    ):
        pass

    def add_graph(
        self,
        graph: Graph,
        target: URIRef | GraphType.DEFAULT | None = None,
    ):
        pass
```

## Graph Scenarios

### Scenario 1: Default Graph (Start with Triple)
Graph instantiated without context becomes a "default" or contextless graph when the first thing added is a triple.
```python
g = Graph()
g.parse(data="<ex:s1> <ex:p1> <ex:o1> .", format="turtle")
print(list(g.triples()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>')]

print(list(g.quads()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>', None)]

print(g.graph_type)
> default  # graph type is now "default"; any triples or quads added after this have no context
g.parse(data="<ex:s2> <ex:p2> <ex:o2> <ex:graph> .", format="nquads")
print(list(g.triples()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>'), ('<ex:s2>', '<ex:p2>', '<ex:o2>')]

print(list(g.quads()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>', None), ('<ex:s2>', '<ex:p2>', '<ex:o2>', None)]
```

---

### Scenario 2: Named Graph (Start with Quad)
Graph instantiated without context gets context from parsed quad.
Subsequently parsed triples inherit the context.
```python
g = Graph()
g.parse(data="<ex:s2> <ex:p2> <ex:o2> <ex:g2> .", format="nquads")
print(list(g.triples()))
> [('<ex:s2>', '<ex:p2>', '<ex:o2>')]

print(list(g.quads()))
> [('<ex:s2>', '<ex:p2>', '<ex:o2>', '<ex:g2>')]

print(g.graph_type)
> named
g.parse(data="<ex:s3> <ex:p3> <ex:o3> .", format="turtle")
print(list(g.triples()))
> [('<ex:s2>', '<ex:p2>', '<ex:o2>'), ('<ex:s3>', '<ex:p3>', '<ex:o3>')]

print(list(g.quads()))
> [('<ex:s2>', '<ex:p2>', '<ex:o2>', '<ex:g2>'), ('<ex:s3>', '<ex:p3>', '<ex:o3>', '<ex:g2>')]
```

---

### Scenario 3: Named Graph with Identifier
Triples added to graph inherit the context.
```python
g = Graph(identifier="ex:g1")
print(g.graph_type)
> named
g.parse(data="<ex:s1> <ex:p1> <ex:o1> .", format="turtle")
print(list(g.triples()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>')]

print(list(g.quads()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>', '<ex:g1>')]
```

---

### Scenario 4: Add quad to default graph

Context is ignored.
```python
g = Graph(graph_type="default")
g.parse(data="<ex:s1> <ex:p1> <ex:o1> <ex:graph> .", format="nquads")
print(list(g.triples()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>')]

print(list(g.quads()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>', None)]
```

---

## Dataset Scenarios

### Scenario 5: Add a Default Graph to a Dataset
```python
g = Graph(graph_type="default")
g.parse(data="<ex:s1> <ex:p1> <ex:o1> .", format="turtle")

ds = Dataset()
ds.add_graph(g)

print(list(ds.triples()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>')]

print(list(ds.quads()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>', None)]
```

---

### Scenario 6: Add a Named Graph to a Dataset
```python
g = Graph(identifier="ex:g1")
g.parse(data="<ex:s2> <ex:p2> <ex:o2> .", format="turtle")

ds = Dataset()
ds.add_graph(g)

print(list(ds.triples()))
> [('<ex:s2>', '<ex:p2>', '<ex:o2>')]

print(list(ds.quads()))
> [('<ex:s2>', '<ex:p2>', '<ex:o2>', '<ex:g1>')]
```

---

### Scenario 7: Add a Graph to the Default Context
Graph ID of graph being added (if present) is overridden by "target".
```python
g = Graph(identifier="ex:g1")
g.parse(data="<ex:s3> <ex:p3> <ex:o3> .", format="turtle")

ds = Dataset()
ds.add_graph(g, target="default")

print(list(ds.triples()))
> [('<ex:s3>', '<ex:p3>', '<ex:o3>')]

print(list(ds.quads()))
> [('<ex:s3>', '<ex:p3>', '<ex:o3>', None)]
```

---

### Scenario 8: Add Graphs to Dataset changing the graph
Graph ID of graph being added (if present) is overridden by "target".
```python
g = Graph(identifier="ex:g2", graph_type="named")
g.parse(data="<ex:s4> <ex:p4> <ex:o4> .", format="turtle")

ds = Dataset()
ds.add_graph(g, target="ex:newg")

print(list(ds.triples()))
> [('<ex:s4>', '<ex:p4>', '<ex:o4>')]

print(list(ds.quads()))
> [('<ex:s4>', '<ex:p4>', '<ex:o4>', '<ex:newg>')]
```

---

### Scenario 9: Iterate Over Triples with Contexts
```python
g1 = Graph(graph_type="default")
g1.parse(data="<ex:s1> <ex:p1> <ex:o1> .", format="turtle")

g2 = Graph(identifier="ex:g1")
g2.parse(data="<ex:s2> <ex:p2> <ex:o2> .", format="turtle")

ds = Dataset()
ds.add_graph(g1)
ds.add_graph(g2)

print(list(ds.triples()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>'), ('<ex:s2>', '<ex:p2>', '<ex:o2>')]

print(list(ds.triples(context=["NAMED", "DEFAULT"])))  # equivalent to default behaviour when not specifying context 
> [('<ex:s1>', '<ex:p1>', '<ex:o1>'), ('<ex:s2>', '<ex:p2>', '<ex:o2>')]

print(list(ds.triples(context="NAMED")))
> [('<ex:s2>', '<ex:p2>', '<ex:o2>')]

print(list(ds.triples(context="DEFAULT")))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>')]

print(list(ds.triples(context=["DEFAULT", "ex:g2"])))  # ex:g2 is not in the dataset so no data returned from this graph.
> [('<ex:s1>', '<ex:p1>', '<ex:o1>')]
```

---

### Scenario 10: Iterate Over Quads with Contexts
```python
g1 = Graph(graph_type="default")
g1.parse(data="<ex:s1> <ex:p1> <ex:o1> .", format="turtle")

g2 = Graph(identifier="ex:g1")
g2.parse(data="<ex:s2> <ex:p2> <ex:o2> .", format="turtle")

ds = Dataset()
ds.add_graph(g1)
ds.add_graph(g2)

print(list(ds.quads()))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>', None), ('<ex:s2>', '<ex:p2>', '<ex:o2>', '<ex:g1>')]

print(list(ds.quads(context="NAMED")))
> [('<ex:s2>', '<ex:p2>', '<ex:o2>', '<ex:g1>')]

print(list(ds.quads(context="DEFAULT")))
> [('<ex:s1>', '<ex:p1>', '<ex:o1>', None)]

print(list(ds.quads(context=["DEFAULT", "ex:g2"])))  # ex:g2 is not in the dataset so no data returned from this graph.
> [('<ex:s1>', '<ex:p1>', '<ex:o1>', None)]
```

**Reply by namedgraph** [reply]
IMO there's no reason for standalone `Graph` to have an identifier. They only get identifiers in the `Dataset` context, i.e. when they become _named_ graphs.

**Reply by PicouAymeric** [reply]
@namedgraph I agree: [W3C spec](https://www.w3.org/TR/rdf11-concepts/#h2_section-rdf-graph) says:
`An RDF graph is a set of [RDF triples].` 
The graph identifier concept is indeed introduced in the graph dataset definition [RDF dataset definition](https://www.w3.org/TR/rdf11-concepts/#h2_section-dataset). 
It is called "graph name" by the way, but since `Graph names are unique within an RDF dataset` and graph names are used to identify graphs within an RDF dataset, identifier is a better name IMO, but maybe we should use w3c terminology anyway.

I also think that the `Graph` class should represent an RDF graph as defined by W3C, thus it should just be a set of triples: without any identifiers.
The `Dataset` class should represent an RDF dataset as defined by W3C: a default graph + the named graphs (a map with key=graph name and value=a graph, or some equivalent structure).

But that's not a little change: we would need to totally refacto graph.py, and that will produce a lot of breaking changes elsewhere in the code.

Also: my understanding would also be that we should eventually remove the `ConjunctiveGraph` class: when you say:
`it provides a "union graph" which most triplestores support` I am not sure to understand: Are you talking about the default graph of a query's RDF dataset is in some triple store by default the merge of all graphs named and unnamed? what we configure with the `SPARQL_DEFAULT_GRAPH_UNION` flag in RDFlib?

When we talk about RDF dataset, it is important to distinguish between:
* The service’s RDF dataset: Defined for all queries (or update), and all users, this contains all triples and graphs potentially accessible. This defines the whole truth of the service.
* The queries’ RDF datasets: This is defined dynamically for each individual query, based on the query itself, the HTTP request, and the service’s RDF dataset

When an RDFlib user instanciate an RDF dataset it will probably be the service's RDF dataset, the query's RDF dataset will be automatically defined after parsing the query. If the user wants to define his query's RDF dataset he will do it with the FROM/FROM NAMED clauses of the sparql query. If I understand well what you mean by "union graph" then this is defined deep in the query processing and thus we don't necessarilly need a class for that.

**Reply by recalcitrantsupplant** [reply]
> An RDF graph is a set of [RDF triples].  

Sounds sensible to stick to the spec then.
Personally I'll probably use the Dataset class by default, pending some of the API updates, as I'm often working with named graphs - I don't see the Graph class would provide much value for me. That's not an argument to remove or change it - obviously it's widely used already.

Being able to set the (equivalent to) queries' dataset for the "triples" and "quads" methods (and related iteration methods) would be good. So being able to iterate over all triples/quads in a dataset regardless of whether it was instantiated with default_union=True.

This is the current behaviour of these methods and what I would find useful:

For a dataset with three triples, two of which are in graphs, the third in the default graph:
```
s1 p1 o1 g1
s2 p2 o2 g2
s3 p3 o3
```

| Function Signature           | Desired Behavior | Desired Result                                                                                                    | Current Result (default_union ON)                                           | Current Result (default_union OFF) |
|------------------------------|------------------|-------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------|----------------------------------|
| `quads()`                    | Iterator over all quads and triples | (s1,p1,o1,g1)<br>(s2,p2,o2,g2)<br>(s3,p3,o3,None)                                                                 | (s1,p1,o1,g1)<br>(s2,p2,o2,g2)<br>(s3,p3,o3,URIRef("urn:x-rdflib:default")) | (s1,p1,o1,g1)<br>(s2,p2,o2,g2)<br>(s3,p3,o3,URIRef("urn:x-rdflib:default")) |
| `triples()`                  | Iterator over dataset triples | <br>(s1,p1,o1)<br>(s2,p2,o2)<br>(s3,p3,o3) | (s1,p1,o1)<br>(s2,p2,o2)<br>(s3,p3,o3)                                      | (s3,p3,o3) |
| `triples(context="g2")`      | Iterator over g2 triples | (s2,p2,o2)                                                                                                        | (s2,p2,o2)                                                                  | (s2,p2,o2) |
| `triples(context="default")`  | Iterator over default graph triples | (s3,p3,o3)                                                                                                        | triples((None, None, None), context=URIRef('urn:x-rdflib:default') returns:<br>(s1,p1,o1)<br>(s2,p2,o2)<br>(s3,p3,o3)                                      | triples((None, None, None), context=URIRef('urn:x-rdflib:default') returns:(s3,p3,o3) |
| `triples(context="named")`   | Iterator over named graphs triples | (s1,p1,o1)<br>(s2,p2,o2)                                                                                          | Not implemented                                                             | Not implemented |
| `quads(context="g2")`        | Iterator over g2 quads | (s2,p2,o2,g2)                                                                                                     | quads((None, None, None, "g2")) returns:<br>(s2,p2,o2,g2)                      | quads((None, None, None, "g2")) returns:<br>(s2,p2,o2,g2) |
| `quads(context="default")`   | Iterator over default graph quads | (s3,p3,o3,None)                                                                                                   | quads((None, None, None, URIRef("urn:x-rdflib:default"))) returns:<br>(s3,p3,o3,URIRef("urn:x-rdflib:default"))                  | quads((None, None, None,URIRef("urn:x-rdflib:default"))) returns:<br>(s3,p3,o3,URIRef("urn:x-rdflib:default")) | (s3,p3,o3,URIRef("urn:x-rdflib:default")) |
| `quads(context="named")`     | Iterator over named graphs quads | (s1,p1,o1,g1)<br>(s2,p2,o2,g2)                                                                                    | Not implemented                                                             | Not implemented |

So this would be behaviour where all triples in the default graph and named graphs are returned unless otherwise specified with context=. This is "inclusive" dataset behaviour under the "Inclusive and Exclusive Datasets" section in this blog: https://blog.metaphacts.com/the-default-graph-demystified

**Reply by namedgraph** [reply]
@PicouAymeric I think `Graph.identifier` is an artifact of the default storage. It makes sense to store only quads, in which case the default graph is just a special named graph with a special identifier. But that is an implementation detail that should not be part of the `Graph` interface and leak into in-memory implementations.

I _think_ we had a similar discussion with @afs about Jena's implementation but I cannot find it now :/

**Reply by PicouAymeric** [reply]
@recalcitrantsupplant I know it would be hard to align with W3C concepts but I really think it is worthy, it will bring a lot of clarity. Right now having a `Graph` class that doesn't represent `RDF Graph` concept and an `Dataset` class that doesn't represent an` RDF dataset` concept is quite confusing, for me at least.

According to the [W3C SPARQL spec](https://www.w3.org/TR/rdf-sparql-query/#rdfDataset) a query should not be executed against a graph, but rather against the representation of a collection of graph that respect the RDF dataset structure. So theoratically the `Graph` class shouldn't have a `query` function at all, only the `Dataset` should have. In order not to break almost everyone's code we could implement the `query` function on `Graph` as follow: query a `Dataset` which default graph is self and doesn't have named graph.

About the `quads` and `triples` functions, I understand the logic of getting everything unless we mention a graph, but if I understand well that's not exactly what metaphacts calls an inclusive dataset: they defined an inclusive dataset as queries' implicit dataset where the default graph is an RDF merge of all graphs named and unnamed, so in that case triples(context="default") could also return all the named graph triples. You consider that `triples(context="default")` should be equivalent to
```
SELECT ?s ?p ?o
{
    GRAPH <urn:defautlgraphURI> {
        ?s ?p ?o 
    }
}
```
But I would be find it more logic to have: `triples(context="default")` equivalent to:
```
SELECT ?s ?p ?o
{
    ?s ?p ?o 
}
```
if you do not have a `GRAPH` clause you are querying the default graph of the query's dataset, which in case of an inclusive dataset should return all the triples of all graph named and unnamed.
By considering `triples(context="default")` should only return one triple you are talking about the service's Dataset default graph or the default graph of a query with an exclusive dataset.


Just to consider all the options: we could also consider that `triples` and `quads` should iterate on the queries' implicit dataset (RDF dataset on which queries without FROM nor FROM NAMED clauses are executed against). However that depends on how we choose to construct the queries' implicit dataset from the service dataset: 
* The default_union ON/OFF, as you mention, which means: should the queries' implicit default graph contain all service named graphs (default_union ON or Inclusive) or only service default graph (default_union OFF or Exclusive)
* Should the service default graph be included in the queries implicit named graphs (if no quads should not return `(s3,p3,o3,something)` at all), if yes should the graph name be UNDEF (or None) or should it be the default graph technical name?
* If we consider default_union ON, do we deduplicate triples when we merge all graphs?

We could also make them iterate on the service's RDF dataset, or let them take from and from_named parameters to iterate on an explicit dataset.

@namedgraph Yes, exactly, I didn't study rdflib storage, but indeed that shouldn't leak into `Graph`: We should set the graph names only in the dataset.
This brings us to another question: Do we implement a parse function for graph or only for dataset? In case we keep the graph parse function, if we have for example a trig file with graph names do we ignore the graph and load all the triples into the graph or do we throw an error? I would say we should keep parse since it is widely used but throw an error.

**Reply by namedgraph** [reply]
I think backwards-compatibility is a major consideration here. There's a lot code depending on these wonky APIs.

**Reply by ashleysommer** [reply]
@recalcitrantsupplant 
> As such I don't expect what I've written below would be backwards compatible 

@PicouAymeric 
> But that's not a little change: we would need to totally refactor graph.py, and that will produce a lot of breaking changes elsewhere in the code.

@namedgraph 
> I think backwards-compatibility is a major consideration here. There's a lot code depending on these wonky APIs.

This discussion is specifically about introducing breaking changes. That is the point of it. The Graph/Dataset interface and behaviour hasn't been changed in the last 15 years, to avoid breaking changes. Now is our chance to fix that.

The RDFLib v8 release is all about breaking changes. There is going to be a new  streaming parser/serializer interface (so all parsers and serializers will face breaking changes and will need to be rewritten), we're getting a new URIRef/BNode implementation that is more performant, but introduces breaking changes. We're completely removing `ConjunctiveGraph`, so every user's code that relies on that, all test suites that import `ConjunctiveGraph` will break, and need to be rewritten. The current `Dataset` implementation is based on a wrapper around `ConjunctiveGraph` so it needs to be re-imagined as well.

Since the start of the RDFLib project in 2001, we've never had a chance like this to introduce the required breaking changes to get things "right". After the release of RDFLib v8, we may never again get the chance.

**Reply by nicholascar** [reply]
Confirming what @ashleysommer says: this is about fixing things not patching. Yes things will break. For people unable to adopt v8, v7 will still be available for quite some time, as v4.22, v5 & v6 still are.

**Reply by recalcitrantsupplant** [reply]
Thanks for the feedback.

To clarify a few points around my thinking, the terminology I'm using is:

1. Inclusive Dataset
- covers all triples/quads in the Default Graph and all Named Graphs
2. Exclusive Dataset 
- graphs targeted must be specified or inferred.
  - default union off: refers to the Default Graph
  - default union on: refers to the union of all Named Graphs.
3. Dataset Default Context
- in SPARQL:
  - set by FROM and FROM NAMED
  - when not set using FROM and FROM NAMED, can be inclusive (e.g. GraphDB) or exclusive (e.g. Fuseki)
- in RDFLib
  - I'm proposing setting via the `context` parameter. It would be an array of one or more of "Named", "Default", and URIs for Graphs.
  - when not set e.g. Graph().triples() or Dataset().quads() I'm proposing the behaviour is inclusive. In my experience many people find the exclusive behaviour counter intuitive.
4. Default Graph: the "unnamed" graph.

What am I proposing:
The `context` parameter on the triples and quads methods (and all related methods that iterate on the dataset e.g. subject_objects etc. ) is equivalent to Dataset clauses in SPARQL i.e. FROM and FROM NAMED *not* equivalent to Graph clauses.

I think the options are:

| Target Data | Context for Inclusive Dataset     | Context for Exclusive Dataset                                                 |
|------------|-------------------------------------------|-------------------------------------------------------------------------------|
| Everything (Default Graph + all Named Graphs) | Context not set OR `["Default", "Named"]` | `["Default", "Named"]`                                                        |
| Default Graph only | `["Default"]`                             | Default Union = true: `["Default"]`<br>Default Union = false: Context not set |
| All Named Graphs only | `["Named"]`                               | Default Union = true: Context not set. <br>Default Union = false: `["Named"]` |
| Specific Named Graphs | `[uri1, uri2]`                            | `[uri1, uri2]`                                                                |
| Default Graph + specific Named Graphs | `["Default", uri1, uri2]`                 | `["Default", uri1, uri2]`                                                     |

I think Inclusive would be the better option here:
 - no need for default union setting
 - easier for new users to not "lose" their data:
   - with Exclusive dataset, adding a quad to the Dataset when Default Union is off means the quad won't be found when using `triples` or `quads` methods unless context is set to "Named". Inverse is true with adding a triple where Default Union is on. 
   - equivalent scenarios for Inclusive Dataset only occur when a user _explicitly_ sets context= to something that does not include their data.

Perhaps the inclusive/exclusive options are a good place to start as the other methods depend on these:
- are the options as I've drawn them correct? other options?
- what are others' preferences?

**Reply by namedgraph** [reply]
@nicholascar I understand 8.x will contain breaking changes, but I still think they should be kept to a minimum. More specifically, they should be limited to where RDFLib is non-compliant or the API is awkward, but not to rename or redesign the APIs that already make sense, just for the sake of redesigning them.

**Reply by namedgraph** [reply]
@recalcitrantsupplant honestly I don't understand/like your proposal :) Why is `GraphType` and `context` necessary at all? As for dataset, I think `get_graph(identifier)` which retrieves a named `Graph` is much more useful than `quads()`.

**Reply by recalcitrantsupplant** [reply]
Hey Martynas, I'll write this up more clearly soon outside of this discussion, including what I think shouldn't be changed. Given the feedback on Graph specifically, i.e. that having a name or identifier for it is inconsistent with the spec, I'll remove that from proposed changes. 

The thinking with quads and related methods, by which I mean "subjects", "subject_predicates", "subject_objects" etc. Is that there should be a way to provide "dataset subsetting" (equivalent to SPARQL FROM and FROM NAMED, but entirely separate from the query method). In my day to day use of rdflib this would be quite useful. 

The main pain point for me with Dataset otherwise is adding and accessing triples or quads, so I'll include proposed changes here too.

**Comment by afs** [comment]
Observations from afar ...

> The context parameter on the triples and quads methods

Would the dataset (the storage unit) have a default context setting? Otherwise if an app changes, then it might require every API call to be tracked down and changed.

FWIW Fuseki has both modes - union default graph is SPARQL only, and it is a view of the dataset at query time. The usual way is to have a setting on the dataset but it can be set per query execution.


For update, where do new triples go to in an inclusive dataset?

**Reply by recalcitrantsupplant** [reply]
> Would the dataset (the storage unit) have a default context setting? Otherwise if an app changes, then it might require every API call to be tracked down and changed.

> For update, where do new triples go to in an inclusive dataset?

I would propose not having a default context setting both for storage and queries; triples added to the dataset without explicitly stating a graph are stored in the Default Graph, quads to the relevant graph.

To confirm, when you say if an app changes... is this in relation to current use of default contexts; those users would need to refactor their code to explicitly state where to store/query triples with each call?

**Comment by PicouAymeric** [comment]
Trying to summarize, what we have to do:
* Graph and Dataset classes in two different files
* Delete conjonctive graph class (merge with Dataset)
* rename "context" by "named graph" in Dataset
* Add add_graph/remove_graph/has_graph etc on Dataset
* A graph is just a set of triples: graph identifier (to rename graph names) shouldn't be on the graph but rather on the dataset, same for store
* Graph doesn't inherit from Node
* Query always executed against an rdf dataset not a graph
* Dataset: is not a Graph but rather based on following structure= one default graph + None or Dictionary IRI,Graph.

What else?

I had a look and it looks like the refactoring will be complex, in order not to break things we need to understand how most of the project works.
I have some time I could spend on that but probably not enough given I don't know the project enough.

**Reply by namedgraph** [reply]
> Query always executed against an rdf dataset not a graph

Why can't it be both? Jena supports that.

**Reply by namedgraph** [reply]
Otherwise looks like a solid summary!

**Reply by PicouAymeric** [reply]
> > Query always executed against an rdf dataset not a graph
> 
> Why can't it be both? Jena supports that.

In theory we always query a dataset and not a graph. "Query a graph" is a shortcut for "query a dataset with the graph being the default graph and no named graph"
So I suggest we could implement that shorcut, something like:
```
class Graph():
  def query(self, query):
     return Dataset(default=self).query(query)
```
and implement all the query logic on the Dataset class

**Reply by namedgraph** [reply]
I think the difference would be though that a `GRAPH` query pattern would never match anything on a plain `Graph` but could reach named graphs on a `Dataset`.

That's how `Model` and `Dataset` queries work in Jena IIRC - but there's a separate `QueryExecution` class.

**Reply by recalcitrantsupplant** [reply]
> I have some time I could spend on that but probably not enough given I don't know the project enough.

At the end of last year I wrote a similar summary to your own (not yet posted), then on a local branch started updating function signatures/reorganising files to reflect the proposed changes, the intention being this would make things clearer for discussion.

I'll spend some time today/tomorrow and put my branch up as a draft PR.

**Reply by namedgraph** [reply]
I wonder if it would be easier to make RDFLib 8.0 just a wrapper for Oxigraph (given that [Oxrdflib](https://github.com/oxigraph/oxrdflib) already exists) rather than try to refactor this legacy codebase 😅 
SPARQL would perform faster, too. @Tpt WDYT? :)

**Reply by ashleysommer** [reply]
@namedgraph
> I wonder if it would be easier to make RDFLib 8.0 just a wrapper for Oxigraph

I get where where you're coming from, but for the last 25 years RDFLib has been _the pure-python RDF implementation_. The whole point of RDFLib is that is is pure python. If you want to wrap C or Rust or Java code, there are better libraries out there (personally, I just use PyOxigraph for my projects).

Having said that, I am planning to have an official Oxigraph-store backend for RDFlib ready for the RDFlib v8 release (something similar to Oxrdflib), that will be an optional python-"extra" in the same way we have optional `orjson` support.

**Reply by Tpt** [reply]
+1 to @ashleysommer RDFLib is the pure-python RDF implementation. Moving everything on top of pyoxigraph seems a larger work than cleaning up RDFLib internals.

Moreover, while pyoxigraph is much faster to load a RDF file or run SPARQL query, plain RDFLib is significantly faster at creating a RDF term or accessing term values (IRI string...) because of the overhead of converting between Python `str` and Rust `String` so for users who don't care about SPARQL but hand write RDF manipulations in Python, RDFLib is likely to be faster.

>  I am planning to have an official Oxigraph-store backend for RDFlib ready for the RDFlib v8 release

This would be amazing! However, feel free to just take over oxrdflib if you prefer to have a smaller RDFLib core and dogfood the RDFLib `Store` plugin API with pyoxigraph. Feel free to reach out if you need help in any case.

**Reply by recalcitrantsupplant** [reply]
I've created a draft PR here: https://github.com/RDFLib/rdflib/pull/3060

**Reply by namedgraph** [reply]
I've provided some comments on the PR

**Comment by edmondchuc** [comment]
Hi all, I've had a chat with @recalcitrantsupplant and will be taking over this task.

Based on the work in https://github.com/RDFLib/rdflib/pull/3060 and @recalcitrantsupplant's recommendations, I will provide a new PR to show the new Dataset class interface with only the core methods. These core methods correspond to the existing ones, and while verbose, will provide the foundation for higher-level methods to be built upon.

Once we reach consensus on the core methods, if there is a demand for it, we will work on use cases for higher-level convenience methods such as data slicing introduced in https://github.com/RDFLib/rdflib/pull/3060.

This might be a good time to look at other higher-level interfaces such as [clownface.js](https://zazuko.github.io/clownface/#/), which provides concise navigation and manipulation of the dataset using method chaining.

A new dataset module will be created to house the new Dataset class. This will exist alongside the current Dataset class and likely tested in a future 7.x release before replacing it completely in an 8.0 release.

### Some other notes relating to Datasets
If our goal is to ultimately remove ConjunctiveGraph and replace it with Dataset, this PR is a good starting point. https://github.com/RDFLib/rdflib/pull/1814

Open issues relating to datasets: [issues with concept: RDF Dataset tag](https://github.com/RDFLib/rdflib/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22concept%3A%20RDF%20dataset%22)

**Reply by namedgraph** [reply]
@edmondchuc if we need inspiration, I think we should be looking at [Jena's RDF API](https://jena.apache.org/tutorials/rdf_api.html) instead.
I made a PoC based on it some time ago: https://github.com/namedgraph/oxijen/blob/main/oxijen/rdf_model.py

**Reply by edmondchuc** [reply]
## RDFLib 7 Dataset Fix
As you might have noticed, we've released a bunch of fixes and enhancements for version 7 in the last few weeks.

Next, I'm planning to fix a long-standing issue in the Dataset class where `Dataset.add_graph` does not add the graph identifier correctly to the `context`. The following code from pyLODE demonstrates a workaround for this problem:

https://github.com/RDFLib/pyLODE/blob/dc65b3189b4e1412cbc17ca74c0b67e782252075/pylode/profiles/supermodel/dataset.py#L1-L28

If we confirm this is indeed a bug, we should roll the fix into version 7 and continue maintaining it until version 8 is released.

## RDFLib 8 Dataset Work

@namedgraph - I think your suggested graph APIs in your opening post https://github.com/RDFLib/rdflib/discussions/2591#discussion-5619172 makes sense. I will create a stub class soon for feedback.

Probably the most difficult and time consuming part will likely be removing the ConjunctiveGraph class inheritance from Dataset. I'd like to tackle this first and get the tests to pass before working on any new interface changes.

**Reply by edmondchuc** [reply]
Hi @namedgraph, could you please take a look at this PR https://github.com/RDFLib/rdflib/pull/3259? It fixes the Dataset.add_graph issue.

**Reply by edmondchuc** [reply]
A couple of more Dataset related PRs:

- https://github.com/RDFLib/rdflib/pull/3262
- https://github.com/RDFLib/rdflib/pull/3263
- https://github.com/RDFLib/rdflib/pull/3264

**Reply by edmondchuc** [reply]
Adding some notes here for Dataset interface changes for v8:

- Dataset methods: triples and serialize (with any triple format) to dump all triples
- Document clearly how to retrieve triples from the default graph (retrieve the graph using Dataset.default_context, soon to become Dataset.default_graph in v8)
- Make default_union=True the default

**Reply by edmondchuc** [reply]
Some more v7 PRs relating to Datasets:

- https://github.com/RDFLib/rdflib/pull/3265
- https://github.com/RDFLib/rdflib/pull/3267
- https://github.com/RDFLib/rdflib/pull/3268

**Reply by namedgraph** [reply]
@edmondchuc wait, does `default_union=True` do what I think it does? Makes the default graph a union of named graphs?
If so, why are we making this change? It will make the non-standard behaviour the default.

**Reply by edmondchuc** [reply]
From my understanding, the default graph at query time is not defined by SPARQL Query 1.1 and is left up to implementations. The specification does, however, provide two examples of possible dataset behaviours, one of which ([13.1 Examples of Datasets - Example 2](https://www.w3.org/TR/sparql11-query/#ex_2)) describes the new proposed `default_union=True` behaviour.

Personally, I find `default_union=True` more intuitive and less of a gotcha for newcomers, but perhaps that is influenced by my past experience with RDF4J. For context, see https://github.com/RDFLib/rdflib/issues/2944.

That said, I welcome your and others' perspectives - both for and against - on adopting this new Dataset behaviour.

**Reply by namedgraph** [reply]
I think it would be a big breaking change without any convincing justification.
To me it's not intuitive and that is perhaps influenced by my experience with Jena :) Where the union graph has a special name.

But more importantly, I think the spec is quite clear about the dataset definition:
> A SPARQL query is executed against an RDF Dataset which represents a collection of graphs. An RDF Dataset comprises one graph, the default graph, which does not have a name, and zero or more named graphs, where each named graph is identified by an IRI.

https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#rdfDataset

**Comment by afs** [comment]
> Make default_union=True the default

What happens if the app writes to the default graph?

**Reply by edmondchuc** [reply]
Triples are still inserted into the default graph with SPARQL Updates. This PR https://github.com/RDFLib/rdflib/pull/3265 contains a test that demonstrates the behaviour more concretely.

**Reply by afs** [reply]
Does the union include the default graph?

If it is the union of named graphs, and not the default unnamed graph, then, in default configuration, added triples are not visible.

**Reply by edmondchuc** [reply]
Yes, querying the default graph returns the union of all graphs, both named and default.

**Reply by namedgraph** [reply]
I don't understand the purpose of flipping the default. I thought these PRs were about fixing the `Dataset`?

**Comment by kuzeko** [comment]
I am not sure to understand the discussion, is this blocking in any way the resolution  of https://github.com/RDFLib/rdflib/issues/3064?

What is needed to fix that?
